docs/design/llvm_sparse_runtime.md

```diff
@@ -0,0 +1,458 @@
+  auto i = reserve_new_element();
+  return get_element_ptr(i);
+}
+```
+
+So is [`append()`](https://github.com/taichi-dev/taichi/blob/172cab8a57fcfc2d766fe2b7cd40af669dadf326/taichi/runtime/llvm/runtime.cpp#L1586-L1589):
+
+```cpp
+void ListManager::append(void *data_ptr) {
+  auto ptr = allocate();
+  std::memcpy(ptr, data_ptr, element_size);
+}
+```
+
+[`clear()`](https://github.com/taichi-dev/taichi/blob/172cab8a57fcfc2d766fe2b7cd40af669dadf326/taichi/runtime/llvm/runtime.cpp#L472-L474) simply resets `num_elements` to zero, without doing anything to the list contents.
+
+## GC
+
+The GC process for a given SNode is divided into three stages:
+
+1. [`gc_parallel_0`](https://github.com/taichi-dev/taichi/blob/172cab8a57fcfc2d766fe2b7cd40af669dadf326/taichi/runtime/llvm/runtime.cpp#L1600-L1626): Moves the remaining, unused indices in `free_list` to its head.
+
+The code differentiates between the cases where destination and source ranges overlap or not:
+
+```sh
+
+# src and dst range do not overlap:  src=[6, 8), dst=[0, 2)
+0   1   2   3   4   5   6   7   8        0   1   2
++---+---+---+---+---+---+---+---+        +---+---+
+|\\\|\\\|\\\|\\\|\\\|\\\|   |   |  --->  |   |   |
++---+---+---+---+---+---+---+---+        +---+---+
+                        src              dst
+
+# src and dst range overlap: src=[3, 8), dst=[0, 5)
+0   1   2   3   4   5   6   7   8        0   1   2   3   4   5
++---+---+---+---+---+---+---+---+        +---+---+---+---+---+
+|\\\|\\\|\\\|   |   |   |   |   |  --->  |   |   |   |   |   |
++---+---+---+---+---+---+---+---+        +---+---+---+---+---+
+            src                          dst
+```
+
+2. [`gc_parallel_1`](https://github.com/taichi-dev/taichi/blob/172cab8a57fcfc2d766fe2b7cd40af669dadf326/taichi/runtime/llvm/runtime.cpp#L1628-L1640): Does some bookkeeping for `free_list` and `recycled_list`.
+
+```cpp
+void gc_parallel_1(RuntimeContext *context, int snode_id) {
+  LLVMRuntime *runtime = context->runtime;
+  auto allocator = runtime->node_allocators[snode_id];
+  auto free_list = allocator->free_list;
+
+  const i32 num_unused =
+      max_i32(free_list->size() - allocator->free_list_used, 0);
+  free_list->resize(num_unused);
+
+  allocator->free_list_used = 0;
+  // 1
+  allocator->recycle_list_size_backup = allocator->recycled_list->size();
+  // 2
+  allocator->recycled_list->clear();
+}
+```
+
+It is important to understand that this stage must be run on a single thread. This is to avoid any kind of data race on `recycled_list`. During GC, `recycled_list` must be cleared, with all its contents being transferred into `free_list`. If this was done in the third stage (which is run in parallel), it is quite difficult to coordinate the event sequence of clearing and transferring among the GPU threads. As a result, this serial stage is created in order to store the number of elements of `recycled_list` into `recycle_list_size_backup` in advance (1), then clears the list (2).
```

Comment:
```suggestion
It is important to understand that this stage must be run on a single thread. This is to avoid any kind of data race on `recycled_list`. During GC, `recycled_list` must be cleared, with all its contents being transferred into `free_list`. If this was done in the third stage (which is run in parallel), it is quite difficult to coordinate the event sequence of clearing and transferring among the GPU threads. As a result, this serial stage is created in order to store the number of elements of `recycled_list` into `recycle_list_size_backup` in advance (1), then clears the list (2). Fortunately, computation here is light and it doesn't take too much time even running in serial.
```